File: ui-server/types/actions/playlist.ts
import { Episode, Playlist } from 'types/app'

export const RECEIVED_PLAYLISTS = 'RECEIVED_PLAYLISTS'
export const RECEIVED_PLAYLIST_EPISODES = 'RECEIVED_PLAYLIST_EPISODES'

export interface ReceivedPlaylistsAction {
  type: typeof RECEIVED_PLAYLISTS
  userId: string
  playlists: Playlist[]
}

export interface ReceivedPlaylistEpisodesAction {
  type: typeof RECEIVED_PLAYLIST_EPISODES
  playlistId: string
  episodes: Episode[]
}

export type PlaylistActionTypes =
  | ReceivedPlaylistsAction
  | ReceivedPlaylistEpisodesAction

File: api/page/playlist.go
package page

import (
	"net/http"

	"github.com/varmamsp/cello/model"
	"github.com/varmamsp/cello/web"
)

func Playlist(c *web.Context, w http.ResponseWriter, req *http.Request) {
	if c.RequirePlaylistId(); c.Err != nil {
		return
	}

	playlist, err := c.App.GetPlaylist(c.Params.PlaylistId)
	if err != nil {
		c.Err = err
		return
	}

	if !c.App.HasPermissionToViewPlaylist(c.Session, playlist) {
		c.Response.StatusCode = http.StatusBadRequest
		return
	}

	episodes, err := c.App.GetPlaylistEpisodes(playlist)
	if err != nil {
		c.Err = err
		return
	}

	podcasts, err := c.App.GetPodcastsForEpisodes(episodes)
	if err != nil {
		c.Err = err
		return
	}

	if c.Session != nil && c.Session.UserId != 0 {
		if err := c.App.LoadPlaybacks(c.Session.UserId, episodes); err != nil {
			c.Err = err
			return
		}
	}

	c.Response.StatusCode = http.StatusOK
	c.Response.Data = &model.ApiResponseData{
		Playlists: []*model.Playlist{playlist},
		Episodes:  episodes,
		Podcasts:  podcasts,
	}
}
